class_13_parallel_programming/homework/homework.py

```diff
@@ -0,0 +1,291 @@
+
+def get_content(url: str) -> bytes:
+    """Sends GET request to a given url. Logs progress and return
+    response in bytes (empty bytes sequence if request failed)
+
+    :param url: url address that stores the image
+    :return: bytes sequence (zero-length if failure)
+    """
+
+    try:
+        with requests.get(url, timeout=(1.05, 3.05)) as r:
+            is_valid, msg = validate_response(r)
+            logging.info(f"GET: {url}\n{msg}")
+            if is_valid:
+                return r.content
+    except requests.exceptions.RequestException:
+        logging.info(f"GET: {url}\nError occurred, skipping this url")
+
+    return b""
+
+
+# manipulating w/ images
+def resize_save_img(buf: bytes, size: Tuple[int, int], n: int) -> None:
+    """Given bytes, turn it to an image, compress it to required format (`size`)
+    ands saves in current directory in jpeg format.
+
+    :param buf: bytes representing downloaded image
+    :param size: maximum allowed dimensions for output image
+    :param n: serial number used for output file name
+    """
+
+    buf = io.BytesIO(buf)
+    with Image.open(buf) as im:
+        im.thumbnail(size)
+        if im.mode != "RGB":
+            im = im.convert("RGB")
+        im.save(f"thumb-{n}.jpg", "JPEG")
+
+
+# preparing for processing images
+def create_output_folder(path_name: str) -> Path:
+    """Creates (if needed) a directory if needed
+
+    :param path_name: path to directory
+    :return: Path object representing a directory
+    """
+
+    path = Path(path_name)
+    logging.debug("Creating output folder")
+    path.mkdir(parents=True, exist_ok=True)
+    return path
+
+
+def convert_thumb_size_param(size: str) -> Tuple[int, int]:
+    """Given `size` string from cl arguments, turn them into ints
+    (width, height)
+
+    :param size: should be a string looking like NxM where N,M are ints
+    :return: tuple of ints: width and height, if successfully converted. Exit program otherwise
+    """
+    try:
+        width, height = map(int, size.lower().split("x"))
+    except ValueError:
+        logging.error("Invalid `size` parameter provided")
+        sys.exit(1)
+    else:
+        if min(width, height) < 1:
+            logging.error(
+                "Both width and height in `size` parameter should be positive"
+            )
+            sys.exit(1)
+    return width, height
+
+
+def validate_args(args: argparse.Namespace) -> None:
+    """Validate most of provided CL arguments, exit program with logged error
+    if something invalid found. Note, `size` CL parameter is validated separately.
+
+    :param args: namespace from `argparse` module containing provided arguments
+    :return: None
+    """
+
+    # validate file name
+    if not Path(args.urls_file).is_file():
+        logging.error("Invalid input file with urls")
+        sys.exit(1)
+
+    # validate size parameter
+    if not args.size:
+        logging.error("Invalid 'size' parameter provided")
+        sys.exit(1)
```

Comment:
It is not possible to have `not args.size` True in the current code version. You either get a tuple, or get an error. So, this verification is redundant.